### PR TITLE
minimalistic html page

### PR DESCRIPTION
**Pull request by gudgl**
## User story
As a supporter, I want to have a simple HTML page to start, list and delete sessions.
Creating new sessions with Curl is too complicated for normal users.

In other words, we need a simple Web interface.

## Task
1. When accessing the base URL of the plexus server via HTTP Get with a browser that sends a header `Content-Type: "text/html"`,  the server must return a minimalistic HTML page with 
* a form to create a new session
* a list of active session
* the option to delete an active session
1. On form submission, the details of the new session are returned HTML formatted. 
1. Create a simple top navigation with the following links
    * List active sessions
    * [Plexus on GitHub](https://github.com/cloudradar-monitoring/plexus/)
1. If authentication is enabled, a dialogue for entering username and password must come up.

#### Create
* The `session ID` must be pre-filled with UUIDv4, but the user can overwrite.
* The `ttl` is pre-filled with the value `3600`
* The username is pre-filled with the value `admin` but the user can overwrite.
* The password is pre-filled with an 18 byte random string not containing special chars.
#### Session created
* By clicking the button, the session opens on a new tab. If the session is password protected, the username and password is passed over via URL parameters. e.g. `http://username:password@example.com/`
* We show two blocks with a ready-to-use script to start the meshagent.
  * For windows
      ```
      iwr -Uri https://github.com/Ylianst/MeshCentral/raw/master/agents/MeshService-signed.exe -OutFile meshagent.exe
      iwr -Uri {AgentMSH URL} -OutFile meshagent.msh
      .\meshagent.exe connect
      ```
  * For Linux
      ```
       curl -s https://github.com/Ylianst/MeshCentral/raw/master/agents/meshagent_x86-64 -o meshagent
       chmod +x meshagent
       curl -s {AgentMSH URL} -o meshagent.msh
       .\meshagent connect
      ```
## Scope
* To follow the Separation of concerns principle we should not render any HTML on the server. The server is pure API. 
* Use some simple JS instead like jQuery to execute the request and to render JSON into HTML in the browser.
* We will not create a fully-fledged app for the session management. 
* Let's include a basic CSS from the CDN of https://mdbootstrap.com/docs/standard/getting-started/installation/

### CODE REVIEW

File: asset/index.html [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+    <meta http-equiv="x-ua-compatible" content="ie=edge"/>
+    <title>Plexus</title>
+    <!-- MDB icon -->
+    <link rel="icon" href="static/img/mdb-favicon.ico" type="image/x-icon"/>
+    <!-- Font Awesome -->
+    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.15.2/css/all.css"/>
+    <!-- Google Fonts Roboto -->
+    <link rel="stylesheet" href="https://fonts.googleapis.com/css2?family=Roboto:wght@300;400;500;700&display=swap"/>
+    <!-- MDB -->
+    <link rel="stylesheet" href="static/css/mdb.min.css"/>
```
Comment (jmattheis) [inline asset/index.html:16]: The issue says, that the files from mdb5 should be included via a CDN, so there should be no need for them to be included into this project. 

> Let's include a basic CSS from the CDN of https://mdbootstrap.com/docs/standard/getting-started/installation/

File: go.mod [hunk 2]
```diff
@@ -4,10 +4,10 @@ go 1.16
 
 require (
 	github.com/go-chi/chi/v5 v5.0.3
+	github.com/go-chi/cors v1.2.0
```
Comment (jmattheis) [inline go.mod:7]: This dep in used right?

File: asset/index.html [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+                + '</div>'
+            $('#session-details').html(dHTML);
+        }
+
+        function deleteSession(id) {
+            console.log(id)
+            if (confirm('Are you sure you want to delete session ' + id + '?')) {
+                $.ajax({
+                    type: 'DELETE',
+                    url: 'https://localhost:8080/session/' + id,
```
Comment (jmattheis) [inline asset/index.html:153]: Plexus isn't required to run on localhost:8080, see how the fetch is done in the share.html.

File: asset/index.html [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+                + '</div>'
+                + '</div>'
+                + '</div>'
+            $('#session-details').html(dHTML);
+        }
+
+        function deleteSession(id) {
+            console.log(id)
+            if (confirm('Are you sure you want to delete session ' + id + '?')) {
+                $.ajax({
```
Comment (jmattheis) [inline asset/index.html:151]: If no IE11 support is needed (See ticket) then [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) should be used here.

File: asset/index.html [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+            $('#created-session').html('');
+        }
+
+        function clearSessionCreated() {
+            $('#session-details').html('');
+        }
+
+        function getSession(agentMSH, sessionURL) {
+            // console.log(agentMSH)
+            let dHTML = '<h2 class="mt-5 mb-3">Session details</h2>'
```
Comment (jmattheis) [inline asset/index.html:75]: If no IE11 support is needed (See ticket) then [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) should be used here.

File: handler/home.go [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+	"net/http"
+
+	"github.com/cloudradar-monitoring/plexus/api"
+	"github.com/cloudradar-monitoring/plexus/asset"
+)
+
+func (h *Handler) Home(rw http.ResponseWriter, r *http.Request) {
+	err := asset.IndexTemplate.Execute(rw, nil)
+	if err != nil {
+		api.WriteHTMLError(rw, http.StatusInternalServerError, "Unable to process the request, try again later.")
```
Comment (jmattheis) [inline handler/home.go:13]: Maybe add the error to the output, otherwise we never know why it failed.

File: api/error.go [hunk 7]
```diff
@@ -38,6 +38,12 @@ func WriteTextError(rw http.ResponseWriter, code int, message string) {
 	fmt.Fprintln(rw, "Desription:", message)
 }
 
+func WriteHTMLError(rw http.ResponseWriter, code int, message string) {
+	rw.Header().Add("content-type", "text/html")
+	rw.WriteHeader(code)
+	_ = json.NewEncoder(rw).Encode(err(code, message))
```
Comment (jmattheis) [inline api/error.go:44]: This isn't an HTML error, this is just a JSON error with HTML content-type.

**Review by jmattheis — COMMENTED** [review]
I haven't fully reviewed this yet, as I still need answer to the questions in the ticket from Thorsten.

File: asset/index.html (hunk 4 above)
Comment (gudgl) [reply asset/index.html]: I will need some help. I couldn't make POST session to work.

File: asset/asset.go [hunk 8]
```diff
@@ -8,4 +8,7 @@ import (
 //go:embed *.html
 var files embed.FS
 
-var ShareTemplate = template.Must(template.ParseFS(files, "share.html"))
+var (
+	ShareTemplate = template.Must(template.ParseFS(files, "share.html"))
+	IndexTemplate = template.Must(template.ParseFS(files, "index.html"))
```
Comment (jmattheis) [inline asset/asset.go:13]: As the index.html isn't really a template, you can just import it as byte array, then there is no need for the error handling:

asset.go
```go
//go:embed share.html
var files embed.FS

//go:embed index.html
var IndexHTML []byte

var (
	ShareTemplate = template.Must(template.ParseFS(files, "share.html"))
)
```
home.go
```go
func (h *Handler) Home(rw http.ResponseWriter, r *http.Request) {
	rw.Header().Set("content-type", "text/html")
    if _, err := rw.Write(asset.IndexHTML); err != nil {
        // log error to h.log
    }
}
```

File: handler/home.go [hunk 9]
```diff
@@ -17,8 +17,8 @@ const internalErrFormat = `
 `
 
 func (h *Handler) Home(rw http.ResponseWriter, r *http.Request) {
-	err := asset.IndexTemplate.Execute(rw, nil)
-	if err != nil {
+	rw.Header().Set("content-type", "text/html")
+	if _, err := rw.Write(asset.IndexHTML); err != nil {
 		log.Error().Err(err).Msg("unable to execute index.html template")
 		api.WriteHTMLErrorf(rw, http.StatusInternalServerError, internalErrFormat, http.StatusInternalServerError, "Unable to process the request, try again later.")
```
Comment (jmattheis) [inline handler/home.go:23]: This doesn't make sense anymore I guess, when the writing to the response on line 21 fails then writing the error to the response writer will also fail. We can remove all code related to html errors.

Removing this line will also fix the linter for this PR.

File: asset/index.html [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+                                   </label>
+                                   <div class="text-box">
+                                       <div class="text-container">
+                                           <span class="text-icon">
+                                               <i class="far fa-copy" onclick="copy('created-linux-session')"></i>
+                                           </span>
+                                           <textarea class="form-control" id="created-linux-session" rows="4" readonly>
+   curl https://raw.githubusercontent.com/Ylianst/MeshCentral/master/agents/meshagent_x86-64 -O meshagent
+   chmod +x meshagent.msh
+   curl ${data.AgentMSH} -OutFile meshagent.msh
```
Comment (jmattheis) [inline asset/index.html:256]: Outfile doesn't exist in curl. Use
```
curl ${data.AgentMSH} > meshagent.msh 
```

**Review by jmattheis — APPROVED** [review]
One little remark, then you can merge this PR.

File: asset/index.html [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+                                   <div class="text-box">
+                                       <div class="text-container">
+                                           <span class="text-icon">
+                                               <i class="far fa-copy" onclick="copy('created-linux-session')"></i>
+                                           </span>
+                                           <textarea class="form-control" id="created-linux-session" rows="4" readonly>
+   curl https://raw.githubusercontent.com/Ylianst/MeshCentral/master/agents/meshagent_x86-64 -O meshagent
+   chmod +x meshagent.msh
+   curl ${data.AgentMSH} > meshagent.msh
+   .\\meshagent connect</textarea>
```
Comment (jmattheis) [inline asset/index.html:257]: This should be ./meshagent connect

File: asset/index.html [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+                                       Start the session on Linux from the Terminal
+                                   </label>
+                                   <div class="text-box">
+                                       <div class="text-container">
+                                           <span class="text-icon">
+                                               <i class="far fa-copy" onclick="copy('created-linux-session')"></i>
+                                           </span>
+                                           <textarea class="form-control" id="created-linux-session" rows="4" readonly>
+   curl https://raw.githubusercontent.com/Ylianst/MeshCentral/master/agents/meshagent_x86-64 -O meshagent
+   chmod +x meshagent.msh
```
Comment (jmattheis) [inline asset/index.html:255]: the meshagent binary should be made executable not the .msh file.

**Review by jmattheis — APPROVED** [review]

**Review by jmattheis — APPROVED** [review]

**Pull request merged**